### PR TITLE
Enable relay by default in New

### LINKED ISSUE #440: Add relay dialer and listener (not hop) to the default libp2p constructor
Seems like a pretty reasonable thing to do. Thoughts @Stebalien @bigs @raulk @vyzo ?

### PR DESCRIPTION
**Pull request by vyzo**
Closes #440

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -44,8 +44,9 @@ type Config struct {
 	Insecure           bool
 	Protector          pnet.Protector
 
-	Relay     bool
-	RelayOpts []circuit.RelayOpt
+	RelayCustom bool
```
Comment (raulk) [inline config/config.go]: `RelayCustom` seems like an bookkeeping flag we use only to track if the user has set relay options explicitly. Can we make it private?

File: options.go [hunk 2]
```diff
@@ -245,9 +255,15 @@ func NATManager(nm config.NATManagerC) Option {
 // NoListenAddrs will configure libp2p to not listen by default.
 //
 // This will both clear any configured listen addrs and prevent libp2p from
-// applying the default listen address option.
+// applying the default listen address option. It also disables relay, unless the
+// user explicitly specifies with an option, as the transport creates an implicit
+// listen address that would make the node diable through any relay it was connected to.
```
Comment (raulk) [inline options.go]: nit: s/diable/dialable

File: options.go [hunk 3]
```diff
@@ -245,9 +255,15 @@ func NATManager(nm config.NATManagerC) Option {
...
 var NoListenAddrs = func(cfg *Config) error {
 	cfg.ListenAddrs = []ma.Multiaddr{}
+	if !cfg.RelayCustom {
+		cfg.RelayCustom = true
```
Comment (raulk) [inline options.go]: Is it necessary to set `RelayCustom` to true here?

File: config/config.go (hunk 1 above)
Comment (vyzo) [reply config/config.go]: I would have liked  this to, but the readers are in a different package.

File: options.go (hunk 2 above)
Comment (vyzo) [reply options.go]: sure, will fix; typos are bugs of sorts :)

File: options.go (hunk 3 above)
Comment (vyzo) [reply options.go]: yes, otherwise the fallback will set it to default and enable it.

File: config/config.go (hunk 1 above)
Comment (raulk) [reply config/config.go]: 👍

Comment (raulk) [reply config/config.go]: Maybe a docstring would clarify the role of this flag, also stating it's not intended for explicit use.

Comment (Stebalien) [reply config/config.go]: Maybe we should change `Relay` to an enum (ish) with `Enabled`, `Disabled`, and `Default` variants.

Comment (raulk) [reply config/config.go]: I thought about that too, but didn’t propose it because it’d imply breakage.

**Comment by Stebalien** [comment]
So, I'd *like* libp2p to be as silent as possible on start (by default). Unfortunately, go-libp2p-circuit tries to figure out if each peer supports relying which adds some additional chatter (that's probably unnecessary in most cases). Could we add an option to disable this first? That is, add an option to disable:

1. Discovering if our peers are willing to relay.
2. Dialing `/p2p-circuit/ipfs/QmTarget` (no relay specified) addresses.

Really, I've still yet to see a use-case for using a random peer as a relay when dialing. In theory, it could help bypass firewalls and censorship but it's effectively a random walk. It may make sense to have some "discover N relays" function that tries to setup N relay "entry points" so we can announce them, but that's a different issue (and something that will happen periodically, not on every connection event).

File: config/config.go (hunk 1 above)
Comment (Stebalien) [reply config/config.go]: This struct actually has an explicit warning saying that it's not stable. Users should use the *options*, they shouldn't modify the config directly.

Comment (vyzo) [reply config/config.go]: ok, I got started in making it an enum and it's extremely cumbersome. 
Better to keep the RelayCustom contraption.

**Comment by vyzo** [comment]
Then again it could be ok as an _option_ -- but it can't be the default, as it has far reaching implications to break unspecific relay dialing by default.

**Comment by vyzo** [comment]
We could also have an option to specify fixed relays, which can be passed to p2p-circuit as known relay nodes so that we can at least have working unspecific relay address dialing.

**Comment by Stebalien** [comment]
> Or do you want to disable the unspecific relay address dialing entirely?
...
> Then again it could be ok as an option -- but it can't be the default, as it has far reaching implications to break unspecific relay dialing by default.

How is the current behavior of unspecified relay addresses useful outside of testing and demos?

---

Basically, I see exactly two use-cases for unspecified relay addresses:

1. Autorelay. In theory, we could insert an address transformer between the routing layer and the swarm, prepending `/p2p-circuit/` to every address once we have enough relays. However, IMO, relay discovery would likely happen in the autorelay module (e.g., on connect, call `circuit.CanHop(p)`).
2. Testing/demoing relay.

**Comment by vyzo** [comment]
Opened #444 for autorelay module.

**Comment by Stebalien** [comment]
So, we should definitely leave the current behavior as the default in go-ipfs until we have a better relay discovery mechanism (as users may be relying on it). It's just that that doing this by default will affect all users of libp2p itself.

**Comment by vyzo** [comment]
Yeah, if we turn it off right now we will break relay dialing completely in ipfs.
So until we have autorelay ready and deployed, we'll have to keep the discovery (and then probably for a while longer).

But we can certainly have a `circuit.DisableRelayDiscovery` option that libp2p users could pass for quiet operation.

**Comment by vyzo** [comment]
Or maybe it should be the default and modify ipfs to pass `EnableRelayDiscovery` so that it keeps working as it currently does.

**Comment by Stebalien** [comment]
> Or maybe it should be the default and modify ipfs to pass EnableRelayDiscovery so that it keeps working as it currently does.

I'd prefer this one. That way, both IPFS and libp2p continue doing what they've always done. The only difference is that libp2p nodes would, by default, be able to dial/accept relayed connections.

**Pull request merged**